### PR TITLE
713-gradcam

### LINKED ISSUE #713: GradCam
Is there any utility for computing and displaying gradcams?

### LINKED ISSUE #1212: Feature Maps - GradCam
Is there any plan to add feature maps and heat maps to the list of features. 
Issue 713 has asked for GradCam, and you already have developed Occlusion. I was wondering if there is any chance to have gradcam and other feature maps and heat maps available for the next version. 

Thank you.

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #713; fixes #1212 

### Description
visualisations with class activation maps
added CAM/GradCAM/GradCAM++

### Status
**ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/visualize/class_activation_maps.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    def __call__(self, x):
+        logits, acti, grad = self.net(x)
+        acti = acti[0]
+        grad = grad[0]
+        b, c, *spatial = grad.shape
+        grad_ave = grad.view(b, c, -1).mean(2)
+        weights = grad_ave.view(b, c, 1, 1)
+        map = (weights * acti).sum(1)
+        return self.norm_features(map)
```
Comment (ahatamiz) [inline monai/visualize/class_activation_maps.py:106]: Normalization not needed here

File: monai/visualize/class_activation_maps.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+#     disp_size = 10
+#     preprocess = transforms.Compose(
+#         [transforms.Resize(resize_param), transforms.ToTensor(), transforms.Normalize(norm_mean, norm_std)]
+#     )
+#     plt.figure(figsize=(disp_size, disp_size))
+#     for i, file in enumerate(glob.glob("./temp/test_images/*")):
+#         image = PIL.Image.open(file)
+#         h, w, b = np.shape(np.array(image))
+#         img_tensor = preprocess(image).unsqueeze(0).to(device)
+#         cam_img = cam_computer(img_tensor)[0].detach().cpu().numpy()
```
Comment (ahatamiz) [inline monai/visualize/class_activation_maps.py:150]: Normalizing cam_img is preferred here. Something as simple as the following should work:

`cam_img = (cam_img-cam_img.min())/ cam_img.max()`

**Review by ahatamiz — CHANGES_REQUESTED** [review]

**Comment by wyli** [comment]
thanks @ahatamiz I've updated the modules, I'll add a few unit tests. do you have a use case in 3D by any chance?

**Comment by ahatamiz** [comment]
> thanks @ahatamiz I've updated the modules, I'll add a few unit tests. do you have a use case in 3D by any chance?

Great work @wyli !  I don't have a trained 3D use case right now.. but maybe we can use one of the COVID models for classification ?

**Comment by wyli** [comment]
added modules and tests for:
- cam
cam
- gradcam
gradcam
- gradcam++
gradcam++

**Review by Nic-Ma — APPROVED** [review]
Thanks for your great work, this new feature looks good to me.
@ericspod do you have any more comments from a researcher's view?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli , and BTW, I think this is very cool a new feature, could you please help also develop a tutorial notebook for it?

Thanks.

**Comment by wyli** [comment]
> Hi @wyli , and BTW, I think this is very cool a new feature, could you please help also develop a tutorial notebook for it?
> 
> Thanks.

that would be nice but I don't have a good quality 3d model, do you have access to any classification model?

**Comment by Nic-Ma** [comment]
Can we use IXI dataset to train a model? Actually, I am curious which part of the feature map indicates the gender in IXI.
Thanks.

**Comment by wyli** [comment]
> Can we use IXI dataset to train a model? Actually, I am curious which part of the feature map indicates the gender in IXI.
> Thanks.

the training is a short demo with only a few examples, don't think visualisation will make sense..

**Comment by Nic-Ma** [comment]
OK, then I don't have any better dataset for reference. Maybe @ericspod knows more?
Thanks.

**Comment by wyli** [comment]
I create a ticket to track the progress of the demo https://github.com/Project-MONAI/tutorials/issues/76
